### PR TITLE
Fix #16274 - LazyInitializer.EnsureInitialized overload for reference types.

### LINKED ISSUE #16274: Proposal: LazyInitializer.EnsureInitialized<T>(ref T target, ref object syncLock, Func<T> valueFactory)
### Background

System.Threading.LazyInitializer has several EnsureInitialized overloads that make it safe and easy to lazily initialize a field, e.g.
```C#
private static SomethingExpensive s_expensive;
public static SomethingExpensive Expensive => LazyInitializer.EnsureInitialized(ref s_expensive, () => SomethingExpensive());
```

It has several overloads, that can be split into two groups:
- Optimistic concurrency
- Pessimistic concurrency

```C#
public static class LazyInitializer
{
    // Optimistic
    public static T EnsureInitialized<T>(ref T target) where T : class;
    public static T EnsureInitialized<T>(ref T target, System.Func<T> valueFactory) where T : class;

    // Pessimistic
    public static T EnsureInitialized<T>(ref T target, ref bool initialized, ref object syncLock);
    public static T EnsureInitialized<T>(ref T target, ref bool initialized, ref object syncLock, System.Func<T> valueFactory);
}
```

The optimistic concurrency overloads don't use locks, only interlockeds.  This means you don't need a separate lock object and the first access can be faster, but in a contended initialization situation, you may end up with multiple objects getting created (only one of them will be published/used).

The pessimistic concurrency overloads do use locks, with a double-checked locking pattern.  This means you can be guaranteed to only have one object instantiated in the case of contention, but you also need a separate lock object.

For the pessimistic overloads, all of them work with any T, regardless of whether it's a value type or a reference type, and thus in addition to accepting an object, they also require passing a reference to a bool, which indicates whether the field has been initialized or not.  That's great for value types, but often for reference types that additional Boolean field is unnecessary, as whether it's been initialized is covered by whether the field is null or not.

### Proposal:

We should add the following overload:
```C#
public static T EnsureInitialized<T>(ref T target, ref object syncLock, System.Func<T> valueFactory) where T : class;
```
(the same as one of the existing pessimistic overloads, but without the ref bool, and with a class constraint on T)

This will allow EnsureInitialized to be used in situations where you're using a ref type, want double-checked locking, and don't want/need to have an extra bool field to track initialization, since that can be done via null/non-null of the target field itself.

### PR DESCRIPTION
**Pull request by WinCPP**
This PR is for the implementation of new overloads of LazyInitializer.EnsureInitialized for reference types.

@stephentoub @terrajobst @karelz Kindly review.

Depends on PR [#9831](https://github.com/dotnet/coreclr/pull/9831) in CoreClr for compilation to succeed.

Thanks,
Mandar

### CODE REVIEW

**Comment by WinCPP** [comment]
Also please note that I missed creating a new branch in my fork for 16274 and commits ended up into 3862 branch. So you may see lots of commits below but the files changed tab shows only the relevant files. I tried creating a new branch and cherry-picking changes, but I got exhausted with git. Hope that is fine.

Thanks,
Mandar

File: src/System.Runtime/tests/System/LazyTests.cs [hunk 1]
```diff
@@ -578,5 +579,55 @@ private struct LIX
...
+
+            for (int count = 0; count < MaxCount; count++)
+            {
+                threads.Add(Task.Run(() => LazyInitializer.EnsureInitialized(ref s_refInitFunctionLockedObject, ref lockObject, () => ClassWithInitFunction.InitFunction())));
+            }
+            Task.WaitAll(threads.ToArray());
+            Assert.Equal(1, ClassWithInitFunction.CountInitialized);
+        }
+
+        private class ClassBase
```
Comment (WinCPP) [inline src/System.Runtime/tests/System/LazyTests.cs]: I had created this as common base with counters and sleep function to be used for ctor based API overload test case. Now that is removed, but I missed merging this into the derived class...

**Comment by WinCPP** [comment]
@stephentoub @karelz seems like the builds are failing because of new API implementation in coreclr not yet being available on corefx side. Did I file the PR too early? Kindly advise.

**Comment by stephentoub** [comment]
> seems like the builds are failing because of new API implementation in coreclr not yet being available on corefx side

Right, this won't work until the CoreCLR changes are merged and resulting binaries are consumed into CoreFx.

> Did I file the PR too early? Kindly advise.

It's fine.  We can re-run it once the changes are available.  It'd be good to link to that dependency PR from this PR description, though.

As an aside, before you submit PRs, please rebase your branch on top of the latest master, or at least squash it; there are 28 commits in this PR, most of which are merge commits.

**Comment by WinCPP** [comment]
> As an aside, before you submit PRs, please rebase your branch on top of the latest master, or at least squash it; there are 28 commits in this PR, most of which are merge commits.

Actually, when I click 'Sync' button on github UI, it seems to be doing the push to remote, along with pull, and that is how all my commits were pushed to remote. Is it ok in this repository to use 'git push --force-with-lease' to push the rebased commit...?

**Comment by stephentoub** [comment]
> Is it ok in this repository to use 'git push --force-with-lease' to push the rebased commit...?

You don't have push access to the dotnet/corefx repo.  But you can push whatever you want to your WinCPP/corefx repo, and any changes you make to your issue-3862 branch will be reflected in this PR.

**Comment by WinCPP** [comment]
@stephentoub @terrajobst @karelz  I'm not sure what happened, but the squashing that I did, this PR got closed. I have reopened it. Now I see 1 commit with 2 files. Please check and review.

Thanks!

File: src/System.Runtime/tests/System/LazyTests.cs [hunk 2]
```diff
@@ -495,6 +496,23 @@ public static void EnsureInitialized_ComplexRefTypes()
             Assert.Null(e);
             Assert.Equal(1, initCount);
             Assert.Null(LazyInitializer.EnsureInitialized(ref e, ref einit, ref elock, () => { initCount++; return null; }));
+
+            // Func based initialization without tracking boolean (uninitialized).
+            string f = null;
+            object fLock = null;
+            Assert.Equal(strTemplate, LazyInitializer.EnsureInitialized(ref f, ref fLock, () => strTemplate));
+            Assert.Equal(strTemplate, f);
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/LazyTests.cs]: We should also Assert.NotNull(fLock)

File: src/System.Runtime/tests/System/LazyTests.cs [hunk 3]
```diff
@@ -495,6 +496,23 @@ public static void EnsureInitialized_ComplexRefTypes()
...
+
+            // Func based initialization without tracking boolean (already initialized).
+            string g = strTemplate;
+            object gLock = null;
+            Assert.Equal(strTemplate, LazyInitializer.EnsureInitialized(ref g, ref gLock, () => strTemplate + "bar"));
+            Assert.Equal(strTemplate, g);
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/LazyTests.cs]: Here we should Assert.Null(gLock)

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@WinCPP, this is failing because the added API doesn't yet exist in corert.  Two things:
1) Can you port your addition to https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Threading/LazyInitializer.cs ?
2) For this PR to go through, we'll need to suppress the warning:
```
D:\j\workspace\AllConfigurat---c1564d53\Tools\ApiCompat.targets(53,5): error : MembersMustExist : Member 'System.Threading.LazyInitializer.EnsureInitialized<T>(T, System.Object, System.Func<T>)' does not exist in the implementation but it does exist in the contract. [D:\j\workspace\AllConfigurat---c1564d53\src\System.Threading\src\System.Threading.csproj]
D:\j\workspace\AllConfigurat---c1564d53\Tools\ApiCompat.targets(67,5): error : ApiCompat failed for 'D:\j\workspace\AllConfigurat---c1564d53\bin\Windows_NT.AnyCPU.Debug\System.Threading\uapaot\System.Threading.dll' [D:\j\workspace\AllConfigurat---c1564d53\src\System.Threading\src\System.Threading.csproj]
```
That can be done by editing the https://github.com/dotnet/corefx/blob/master/src/System.Threading/src/ApiCompatBaseline.uapaot.txt file; you can use this recent PR as an example:
https://github.com/dotnet/corefx/pull/16561/files

**Comment by stephentoub** [comment]
@WinCPP, why did you close this?

**Comment by WinCPP** [comment]
By mistake... I was trying to squash commit and same that as other day happened... Can you please help reopen it... I don't know what happened...

**Comment by stephentoub** [comment]
I'm unable to because the new commits you added are no longer in your branch, so there's nothing new for the PR to display.
https://github.com/WinCPP/corefx/commits/issue-3862

**Comment by WinCPP** [comment]
Phew... back open... I will do the changes that you mentioned above now...

File: src/System.Runtime/tests/System/LazyTests.cs [hunk 4]
```diff
@@ -463,27 +464,35 @@ public static void EnsureInitialized_ComplexRefTypes()
             object aLock = null;
             Assert.NotNull(LazyInitializer.EnsureInitialized(ref a, ref aInit, ref aLock));
             Assert.NotNull(a);
+            Assert.True(aInit);
+            Assert.NotNull(aLock);
```
Comment (WinCPP) [inline src/System.Runtime/tests/System/LazyTests.cs]: @stephentoub I have added asserts for true / not null in earlier test cases as well. Please let me know if this is ok...

**Comment by karelz** [comment]
@stephentoub for my own education - how is CoreRT API surface/implementation propagated to CoreFX? Is there similar regular ingestion like for CoreCLR?

**Comment by WinCPP** [comment]
@stephentoub I have done the two tasks
1. Porting code to corert: pull request [link](https://github.com/dotnet/corert/pull/2868).
2. Done the modifications to suppress warnings.

Please check.

**Comment by stephentoub** [comment]
> Is there similar regular ingestion like for CoreCLR?

Yes

File: src/System.Runtime/src/ApiCompatBaseline.uapaot.txt [hunk 5]
```diff
@@ -244,5 +244,6 @@ CannotRemoveBaseTypeOrInterface : Type 'System.Threading.WaitHandle' does not in
 CannotRemoveBaseTypeOrInterface : Type 'System.Threading.Tasks.Task' does not implement interface 'System.IDisposable' in the implementation but it does in the contract.
 MembersMustExist : Member 'System.Threading.Tasks.Task.Dispose()' does not exist in the implementation but it does exist in the contract.
 MembersMustExist : Member 'System.Threading.Tasks.Task.Dispose(System.Boolean)' does not exist in the implementation but it does exist in the contract.
+MembersMustExist : Member 'System.Threading.LazyInitializer.EnsureInitialized<T>(T, System.Object, System.Func<T>)'
```
Comment (stephentoub) [inline src/System.Runtime/src/ApiCompatBaseline.uapaot.txt]: This needs to be the full error message, as in the other cases in the file.

Comment (WinCPP) [inline src/System.Runtime/src/ApiCompatBaseline.uapaot.txt]: I'm sorry. While cutting out the trailing csproj in compilation error, I chewed up a bit more. I should have seen other lines. Honestly, been doing silly mistakes over my limit. Need to introspect why.

**Comment by WinCPP** [comment]
@stephentoub Is the above edit sufficient... because those two builds are still failing with same error...

**Comment by stephentoub** [comment]
> Is the above edit sufficient

It's the wrong file.  It needs to be the one in src\System.Threading\ rather than src\System.Runtime

**Comment by WinCPP** [comment]
> It's the wrong file. It needs to be the one in src\System.Threading\ rather than src\System.Runtime
😶 hope it goes through...

**Pull request merged**